pkg/gofr/cmd/terminal/progress.go

```diff
@@ -0,0 +1,120 @@
+package terminal
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+)
+
+var (
+	errTermSize     = errors.New("error getting terminal size, could not initialize progress bar")
+	errInvalidTotal = errors.New("error initializing progress bar, total should be > 0")
+)
+
+type ProgressBar struct {
+	stream  Output
+	current int64
+	total   int64
+	tWidth  int
+	mu      sync.Mutex
+}
+
+type Term interface {
+	IsTerminal(fd int) bool
+	GetSize(fd int) (width, height int, err error)
+}
+
+func NewProgressBar(out Output, total int64) (*ProgressBar, error) {
+	p := &ProgressBar{
+		stream:  out,
+		total:   total,
+		tWidth:  0,
+		current: 0,
+	}
+
+	w, _, err := out.getSize()
+	if err != nil {
+		return p, errTermSize
+	}
+
+	if total < 0 {
+		p.total = 0
+
+		return p, errInvalidTotal
+	}
+
+	p.tWidth = w
+
+	return p, nil
+}
+
+func (p *ProgressBar) Incr(i int64) bool {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.current < p.total {
+		p.current += i
+		p.current = min(p.current, p.total)
```

Comment:
current could be a sync/atomic.Int64

It would avoid the mutex and the defer

https://pkg.go.dev/sync/atomic#Int64

But the min thing might need to be rewritten, maybe when reading the value

```suggestion
		p.current.Add(i)
		p.current = min(p.current, p.total)
```

Reply:
But we also want to synchronize the painting of the progress bar that is being called in the updateProgressBar(), so mutex is either way mutex may be required.

Reply:
Oh that's right. It's a valuable comment you could add to the code

Reply:
Added.